torch/jit.py

```diff
@@ -1,26 +1,375 @@
+# - It is given an underlying function, which knows how to actually
+#   execute the code that we want to compile.
+# - When we encounter an input configuration for which we don't
+#   have an optimized trace, we run the underlying function, tracing its
+#   result.  The trace is not done yet, so we save it into our set of pending
+#   traces for that configuration.
+# - When we encounter an input configuration whose trace is "ready"
+#   (that is, we've seen all of the passes, so the trace contains
+#   forwards/backwards/etc), we compile it, and then register this
+#   as the compiled trace.
+# - When we encounter an input configuration whose trace is compiled,
+#   we just directly run the compiled trace.
+class CompiledModule(Module):
+    def __init__(self, inner, params=tuple(), **kwargs):
+        super(CompiledModule, self).__init__()
+        self.inner = inner
+        self.params = ParameterList(list(params))
+        self.kwargs = kwargs
+        self.ktrace_cache = {}
+
+    def _process_args(self, args):
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        # TODO: I'm not entirely sure about volatile implies nderivs=0 special case
+        is_volatile, in_vars_key = vars_key(in_vars)
+        in_key = (in_vars_key, in_struct)
+        return in_vars, in_struct, is_volatile, in_key
+
+    # NB: In principle, there could also be a 'raw' version of this compiler,
+    # but since the logic is so complicated, testing code wouldn't benefit much
+    def forward(self, *args):
+        in_vars, in_struct, is_volatile, in_key = self._process_args(args)
+        ktrace = self.ktrace_cache.get(in_key)
+        if ktrace is None:
+            ktrace = TraceForKey(self.inner, volatile=is_volatile, **self.kwargs)
+            self.ktrace_cache[in_key] = ktrace
+        # TODO: avoid virtual method call here
+        closure = ktrace.maybe_closure()
+        if closure:
+            # We already compiled it!  Run it directly, and
+            # use the saved out_struct to unflatten.
+            out_vars = closure()(*in_vars)
+            out_struct = ktrace.out_struct
+        else:
+            # No compiled trace available.  Run it by hand.
+            out_vars, out_struct = ktrace.add_trace(args, in_vars, in_struct)
+        if isinstance(out_vars, Variable):
+            out_vars = (out_vars, )
+        out, extras = _unflatten(out_vars, out_struct)
+        assert len(extras) == 0
+        return out
+
+    def has_trace_for(self, *args):
+        in_vars, in_struct, is_volatile, in_key = self._process_args(args)
+        ktrace = self.ktrace_cache.get(in_key)
+        if ktrace is None:
+            return False
+        return ktrace.maybe_closure() is not None
+
+    # TODO: Provide more compiled code management utility methods
+
+
+# CompiledModule memoizes multiple traces and switches between them based on
+# inputs provided to a call; a TraceForKey represents one such trace. Things
```

Comment:
This doesn't seem right. TraceForKey actually contains multiple traces, all of which share input configuration

Reply:
Yeah, this is a bit subtle. I added some more text.